File: android/app/src/main/java/com/kordar/ghostchat/core/managers/ConnectionFileTransferRouter.kt
package com.kordar.ghostchat.core.managers

import com.kordar.ghostchat.core.files.ChunkTimeoutTracker
import com.kordar.ghostchat.core.files.FileTransferService
import com.kordar.ghostchat.models.ControlMessage
import kotlinx.coroutines.flow.MutableSharedFlow

/**
 * File-transfer control-message routing for [ConnectionManager]. Extracted so the
 * manager itself stays under the 400-LOC cap — this handler owns no state; the
 * [FileTransferService], timeout tracker, and outbound flows all live on the manager.
 */
internal class ConnectionFileTransferRouter(
    private val fileTransfer: () -> FileTransferService,
    private val chunkTimeout: ChunkTimeoutTracker,
    private val incomingFile: MutableSharedFlow<FileTransferService.IncomingFile>,
    private val fileTransferAborted: MutableSharedFlow<String>,
    private val sendControl: suspend (ControlMessage) -> Unit,
    private val awaitSendSlot: suspend () -> Unit
) {

    suspend fun route(ctrl: ControlMessage) {
        val ft = fileTransfer()
        when (ctrl) {
            is ControlMessage.FileStart -> {
                ft.handleStart(
                    fileId = ctrl.fileId, name = ctrl.name, size = ctrl.size,
                    mimeType = ctrl.mimeType, totalChunks = ctrl.totalChunks
                )
                chunkTimeout.arm(ctrl.fileId)
            }

            is ControlMessage.FileChunk -> {
                ft.handleChunk(
                    fileId = ctrl.fileId, index = ctrl.index, base64Data = ctrl.data
                )
                chunkTimeout.progressed(ctrl.fileId)
            }

            is ControlMessage.FileComplete -> {
                val ev = ft.handleComplete(ctrl.fileId, ctrl.sha256)
                when (ev) {
                    is FileTransferService.Event.Completed -> {
                        chunkTimeout.cancel(ctrl.fileId)
                        incomingFile.tryEmit(ev.file)
                    }
                    is FileTransferService.Event.Missing ->
                        runCatching { sendControl(ControlMessage.FileRetransmit(ctrl.fileId, ev.indices)) }
                    is FileTransferService.Event.IntegrityFailure -> {
                        chunkTimeout.cancel(ctrl.fileId)
                        fileTransferAborted.tryEmit(ctrl.fileId)
                    }
                    else -> Unit
                }
            }

            is ControlMessage.FileRetransmit -> {
                val msgs = ft.retransmitMessages(ctrl.fileId, ctrl.indices)
                for (m in msgs) { awaitSendSlot(); runCatching { sendControl(m) } }
            }

            else -> Unit
        }
    }
}

File: ios/GhostChat/Core/Managers/ConnectionFileTransferRouter.swift
import Foundation

/// File-transfer control-message routing for `ConnectionManager`. Extracted so the
/// manager itself stays under the 400-LOC cap — this handler owns no state (the
/// `fileTransfer` service and the timeout tracker live on the manager), it just
/// pattern-matches `ControlMessage` variants and calls back into the caller.
@MainActor
extension ConnectionManager {

    func routeFileControl(_ ctrl: ControlMessage) async {
        switch ctrl {
        case .fileStart(let fileId, let name, let size, let mimeType, let totalChunks):
            _ = fileTransfer.handleStart(
                fileId: fileId, name: name, size: size,
                mimeType: mimeType, totalChunks: totalChunks
            )
            chunkTimeout.arm(fileId: fileId)

        case .fileChunk(let fileId, let index, let data):
            _ = fileTransfer.handleChunk(fileId: fileId, index: index, base64Data: data)
            chunkTimeout.progressed(fileId: fileId)

        case .fileComplete(let fileId, let sha256):
            let event = fileTransfer.handleComplete(fileId: fileId, expectedSha256Hex: sha256)
            switch event {
            case .completed(let file):
                chunkTimeout.cancel(fileId: fileId)
                fileContinuation?.yield(file)
            case .missing(_, let indices):
                try? await sendControl(.fileRetransmit(fileId: fileId, indices: indices))
            case .integrityFailure:
                chunkTimeout.cancel(fileId: fileId)
                abortContinuation?.yield(fileId)
            default:
                break
            }

        case .fileRetransmit(let fileId, let indices):
            let messages = fileTransfer.retransmitMessages(fileId: fileId, indices: indices)
            for msg in messages {
                try? await awaitSendSlotForRoute()
                try? await sendControl(msg)
            }

        case .pushToken, .notifyToken:
            // Peer-supplied push tokens — surface to subscribers so save-contact
            // flows can persist them. P0 #1 wiring.
            handlePeerToken(ctrl)

        case .renegotiate, .callRequest, .callResponse, .callEnd,
             .securityAlert, .messageAck, .messageRead, .ready,
             .typing, .capabilities,
             .messageDelete, .messageEdit, .messagePin:
            // Other control types are handled in their own phases — ignore here so they
            // don't get mistaken for text messages.
            break
        }
    }
}
